src/imp.rs

```diff
@@ -0,0 +1,288 @@
+        TreePtr {
+            inner: N::from_repr(node) as *const N,
+        }
+    }
+
+    /// Casts this ptr across equivalent reprs.
+    pub fn cast<U>(this: TreePtr<T, N>) -> TreePtr<T, U>
+    where
+        U: TransparentNewType<Repr = SyntaxNode<T>>,
+    {
+        // We can avoid an Arc bump if we want here, but lets leverage existing
+        // `new`, to minimize unsafety.
+        let r: &SyntaxNode<T> = this.into_repr();
+        let n = U::from_repr(r);
+        TreePtr::new(n)
+    }
+}
+
+impl<T, N> Drop for TreePtr<T, N>
+where
+    T: Types,
+    N: TransparentNewType<Repr = SyntaxNode<T>>,
+{
+    // Drop refcount
+    fn drop(&mut self) {
+        // Be careful not to leave dangling references to node.
+        let root: *const SyntaxRoot<T> = {
+            let node: &N = &*self;
+            let node = node.into_repr();
+            node.root
+        };
+        drop(unsafe { Arc::from_raw(root) });
+        // inner may be a dangling pointer at this point, but it is okey: it's a
+        // raw pointer after all
+    }
+}
+
+// Manual impls due to `inner: *const N` field
+unsafe impl<T, N> Send for TreePtr<T, N>
+where
+    T: Types,
+    N: TransparentNewType<Repr = SyntaxNode<T>>,
+    N: Send,
+{
+}
+
+unsafe impl<T: Sync, N> Sync for TreePtr<T, N>
+where
+    T: Types,
+    N: TransparentNewType<Repr = SyntaxNode<T>>,
+    N: Sync,
+{
+}
+
+impl<T, N> std::ops::Deref for TreePtr<T, N>
+where
+    T: Types,
+    N: TransparentNewType<Repr = SyntaxNode<T>>,
+{
+    type Target = N;
+    fn deref(&self) -> &N {
+        // We are a `TreePtr`, so underlying `SyntaxRoot` has an count at least
+        // one, so it's safe to deref a node.
+        unsafe { &*self.inner }
+    }
+}
+
+impl<T: Types> SyntaxNode<T> {
+    pub(crate) fn new_root(green: GreenNode<T>, data: T::RootData) -> TreePtr<T, SyntaxNode<T>> {
+        let mut root = Arc::new(SyntaxRoot {
+            root_node: SyntaxNode::new_impl(ptr::null(), None, green),
+            data,
+        });
+
+        let red_node: *mut SyntaxNode<T> = &mut Arc::get_mut(&mut root).unwrap().root_node;
+        // "forget" the `root` so that we have rc equal to one once exiting this
+        // function.
+        let root_ptr: *const SyntaxRoot<T> = Arc::into_raw(root);
+        // set backreference
+        unsafe {
+            (*red_node).root = root_ptr;
+        }
+        TreePtr { inner: red_node }
+    }
+
+    fn new_child(
+        &self,
+        start_offset: TextUnit,
+        index_in_parent: usize,
+        green: GreenNode<T>,
+    ) -> SyntaxNode<T> {
+        let parent_data = ParentData {
+            parent: ptr::NonNull::from(self),
+            start_offset,
+            index_in_parent: index_in_parent as u32,
+        };
+        SyntaxNode::new_impl(self.root, Some(parent_data), green)
+    }
+
+    fn new_impl(
+        root: *const SyntaxRoot<T>,
+        parent: Option<ParentData<T>>,
+        green: GreenNode<T>,
+    ) -> SyntaxNode<T> {
```

Comment:
Considering that (if I understand correctly) letting this `SyntaxNode` live longer than the SyntaxRoot pointed at by `root` would lead to a dangling pointer, I wonder if this should be marked unsafe... Not sure whether that's idiomatic :)